### PR TITLE
samples: subsys: ipc: basic: Basic cores communication

### PR DESCRIPTION
**Pull request by jaz1-nordic**
Add basic cores communication sample to show how to send messages between two cores using mbox and shared data structure.

### CODE REVIEW

File: samples/subsys/ipc/basic/README.rst [hunk 1]
```diff
@@ -0,0 +1,48 @@
+.. zephyr:code-sample:: ipc-basic
+   :name: IPC service: basic struct
+   :relevant-api: ipc
+
+   Send messages between two cores using mbox and shared data structure.
+
+Overview
+********
+
+This application demonstrates how to communcate between cores without any backend in
```
Comment (magp-nordic) [inline samples/subsys/ipc/basic/README.rst:10]:
```suggestion
This application demonstrates how to communicate between cores without any backend in
```

File: samples/subsys/ipc/basic/README.rst [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+- Stop bits: 1
+
+Reset the board and the following message will appear on the corresponding
+serial port, one is host another is remote:
+
+.. code-block:: console
+
+   *** Booting Zephyr OS build v3.7.0-rc1-427-g7f891a6a40e8 ***
+   I: No data from remote
+   I: Sent 14 bytes to host
```
Comment (magp-nordic) [inline samples/subsys/ipc/basic/README.rst:38]:
```suggestion
   I: Sent 14 bytes to remote
```
I guess it should be remote, as it is the host's log.

File: samples/subsys/ipc/basic/sample.yaml [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+    extra_args:
+      basic_SNIPPET=nordic-flpr
+    sysbuild: true
+    harness: console
+    harness_config:
+      type: multi_line
+      ordered: false
+      regex:
+        - "No data from remote"
+        - "Sent 14 bytes to host"
```
Comment (magp-nordic) [inline samples/subsys/ipc/basic/sample.yaml:18]:
```suggestion
        - "Sent 14 bytes to remote"
```

File: samples/subsys/ipc/basic/src/main.c [hunk 4]
```diff
@@ -0,0 +1,194 @@
...
+
+static const struct mbox_dt_spec rx_channel = MBOX_DT_SPEC_GET(DT_PATH(mbox_consumer), rx);
+static const struct mbox_dt_spec tx_channel = MBOX_DT_SPEC_GET(DT_PATH(mbox_consumer), tx);
+static shared_t *rx_data = (shared_t *)((uint8_t *)(DT_REG_ADDR(DT_NODELABEL(sram_rx))));
+static shared_t *tx_data = (shared_t *)((uint8_t *)(DT_REG_ADDR(DT_NODELABEL(sram_tx))));
+
+data_packet_t data_packet;
+
+/**
+ * @brief Send data to aother core
```
Comment (magp-nordic) [inline samples/subsys/ipc/basic/src/main.c:26]:
```suggestion
 * @brief Send data to another core
```

File: samples/subsys/ipc/basic/src/main.c [hunk 5]
```diff
@@ -0,0 +1,194 @@
...
+	}
+
+	return 0;
+}
+
+int main(void)
+{
+	int bytes_sent = 0;
+
+	int ret = mbox_init();
```
Comment (magp-nordic) [inline samples/subsys/ipc/basic/src/main.c:159]:
```suggestion
	int bytes_sent = 0;
	int ret = mbox_init();
	
```
To fix compliance check

File: samples/subsys/ipc/basic/src/main.c [hunk 6]
```diff
@@ -0,0 +1,194 @@
...
+	/* Clear lock */
+	atomic_flag_clear(&rx_data->lock);
+
+	/* Return how many items were read */
+	return size;
+}
+
+/**
+ * @brief Verify whether another core has data ready for us to read
+ * @return -1 if lock is not acquired, otherwise how many items are available for reading
```
Comment (magp-nordic) [inline samples/subsys/ipc/basic/src/main.c:95]: I would put it the other way around:
"Number of items available for reading, -1 if lock is not acquired"

File: samples/subsys/ipc/basic/remote/src/main.c [hunk 7]
```diff
@@ -0,0 +1,194 @@
...
+static shared_t *rx_data = (shared_t *)((uint8_t *)(DT_REG_ADDR(DT_NODELABEL(sram_rx))));
+static shared_t *tx_data = (shared_t *)((uint8_t *)(DT_REG_ADDR(DT_NODELABEL(sram_tx))));
+
+data_packet_t data_packet;
+
+/**
+ * @brief Send data to aother core
+ * @param buffer
+ * @param size
+ * @return -1 if lock is not acquired, otherwise how many items were transfered
```
Comment (magp-nordic) [inline samples/subsys/ipc/basic/remote/src/main.c:29]: Same as in the host main.c

File: samples/subsys/ipc/basic/remote/src/main.c [hunk 8]
```diff
@@ -0,0 +1,194 @@
...
+ */
+int send_to_host(const unsigned char *const restrict buffer, const unsigned int size)
+{
+	/* Try and get lock */
+	if (atomic_flag_test_and_set(&tx_data->lock)) {
+		/* Return -1 in case lock is not acquired (used by other core)*/
+		return -1;
+	}
+
+	/* Limit size of buffer to be transfered to the size of destine buffer */
```
Comment (magp-nordic) [inline samples/subsys/ipc/basic/remote/src/main.c:39]: I guess it should be "destination buffer" or "destined buffer".

File: samples/subsys/ipc/basic/remote/src/main.c [hunk 9]
```diff
@@ -0,0 +1,194 @@
...
+
+	/* Return how many items were transfered */
+	return tx_data->size;
+}
+
+/**
+ * @brief Get data from another core
+ * @param buffer
+ * @param size
+ * @return -1 if lock is not acquired, otherwise how many items were read
```
Comment (magp-nordic) [inline samples/subsys/ipc/basic/remote/src/main.c:63]: Same as in the host main.c

**Review by magp-nordic — APPROVED** [review]
I left a few nitpicking notes. I am also wondering: is it possible that one of the first mbox messages ("No data from host/remote") is missed because the other core enabled mbox after some delay? Or maybe mbox does not send anything until it is enabled by both cores?
Besides that, LGTM.

**Comment by jaz1-nordic** [comment]
Changes moved to https://github.com/nrfconnect/sdk-zephyr/pull/1910 .

**Pull request closed without merging**